Add tracker variable name in the coverage object

lpinca opened the pull request:
I have a special use case where I need to know the name of the tracking variable for a given coverage object.
To do this I can get the md5 of the [`path`](https://github.com/gotwarlost/istanbul/blob/4799c5823da130753cbe0f1c52423be6b9fb39dd/lib/instrumenter.js#L518) property like you do [here](https://github.com/gotwarlost/istanbul/blob/4799c5823da130753cbe0f1c52423be6b9fb39dd/lib/instrumenter.js#L43-L59) but to make it easier this patch exposes the variable name in a `trackerVar` property.

gotwarlost commented:
Why do you need this? The other way to have a well-known tracker variable name is to set the option to omit the tracker suffix.

I'm not inclined to expose the internals of istanbul's processing which is an implementation detail in the API. I'm actually in the middle of some large changes and this stuff might actually change.

lpinca commented:
Ok, I understand.
I'll try to explain the reason that led me to this hack.
In [Primus](https://github.com/primus/primus) we have a plugin system where the client part of a plugin gets [stringified](https://github.com/primus/primus/blob/80718481d852fc99ed712ecc19c914f0a85f92ad/index.js#L578) and added to the client library (this is an [example](https://github.com/primus/primus-emit/blob/61f6e0200fe44560671cd6bb7a92d035e59f4788/index.js#L9-L43) of what the client side of a plugin looks like).
After this we [build](https://github.com/primus/primus/blob/80718481d852fc99ed712ecc19c914f0a85f92ad/index.js#L170-L177) the client in a new context.

When using istanbul the plugin module gets instrumented but when the client part is stringified and added to the main client it no longer have access to the tracking variable because it is only defined in the module context.

With the above approach I was able to inject the tracking variable as a global property in the context where the client gets built, fixing the issue.

Do you have any suggestion?

gotwarlost commented:
You want to inject the whole coverage object into the execution scope, right? The tracker variable is actually re-assigned a couple of times so you want the full coverage object.

See https://github.com/gotwarlost/istanbul/pull/294 for similar work to make `vm.runInContext` do something similar.

lpinca commented:
I don't think that the work in #294 or #465 will help in this case.
To better explain the issue I've created a test case in this [gist](https://gist.github.com/lpinca/7f5cac993da4143ac69c).

The problem is [here](https://gist.github.com/lpinca/7f5cac993da4143ac69c#file-test-js-L9). Basically we get some instrumented code and use it in a place where the tracking variable is not defined. Instrumenting the code in the `hook.runInContext` and injecting the full coverage object will not help.

If you have few minutes please take a look at it. I would really like to find a workaround for this.

Thank you.

gotwarlost commented:
Thanks for the explanation. I see the problem now. 

Are you looking for istanbul to not generate instrumentation for these functions or are you actually trying to get them instrumented as well? I don't think the latter will work if you are shipping these things on the wire.

lpinca commented:
We would like to get the coverage of those functions as well, otherwise we only have a partial coverage of the plugin code but I understand that there is no easy way.

Injecting the tracking variables as global properties in the new context was the only way I found to make it work.

```js
context.__cov_whatever = global.$$cov_1445375721569$$['/foo/bar/filename.js'];
```

That said, having an option to not instrument a part of the code could help. At least the process will not crash.
